### PR TITLE
Misleading error for pd.read_msgpack

### LINKED ISSUE #27160: Misleading error for pd.read_msgpack
#### Code Sample, a copy-pastable example if possible

```python
import pandas as pd
pd.read_msgpack('this/path/does/not/exist')

```
#### Problem description

Such an error is misleading because it suggests that there is a problem with the datatype being passed, not that the path does not exist. The error raised is:

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File ".local/anaconda3/lib/python3.7/site-packages/pandas/io/packers.py", line 226, in read_msgpack
    raise ValueError('path_or_buf needs to be a string file path or file-like')
ValueError: path_or_buf needs to be a string file path or file-like
```

#### Expected Output

Raise an error indicating that the path was not found.

#### Output of ``pd.show_versions()``

<details>
>>> pd.show_versions()

INSTALLED VERSIONS
------------------
commit: None
python: 3.7.3.final.0
python-bits: 64
OS: Linux
OS-release: 4.18.0-24-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.24.2
pytest: 4.3.1
pip: 19.0.3
setuptools: 40.8.0
Cython: 0.29.6
numpy: 1.16.2
scipy: 1.2.1
pyarrow: None
xarray: None
IPython: 7.4.0
sphinx: 1.8.5
patsy: 0.5.1
dateutil: 2.8.0
pytz: 2018.9
blosc: None
bottleneck: 1.2.1
tables: 3.5.1
numexpr: 2.6.9
feather: None
matplotlib: 3.0.3
openpyxl: 2.6.1
xlrd: 1.2.0
xlwt: 1.3.0
xlsxwriter: 1.1.5
lxml.etree: 4.3.2
bs4: 4.7.1
html5lib: 1.0.1
sqlalchemy: 1.3.1
pymysql: None
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datarea

</details>

### PR DESCRIPTION
**Pull request by U09Kane**
- [x] closes #27160 
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/io/packers.py [hunk 1]
```diff
@@ -115,8 +115,12 @@ def writer(fh):
 
     path_or_buf = _stringify_path(path_or_buf)
     if isinstance(path_or_buf, str):
-        with open(path_or_buf, mode) as fh:
-            writer(fh)
+        try:
+            with open(path_or_buf, mode) as fh:
+                writer(fh)
+        except (FileNotFoundError, ValueError):
```
Comment (jreback) [inline pandas/io/packers.py]: why is there a need to catch ValueError here?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/io/packers.py (hunk 1 above)
Comment (U09Kane) [reply pandas/io/packers.py]: I was also thinking that when I included it but I defaulted to leaving it in as a “just to be safe”. 

I’m all for removing it though if it’s unnecessary.

File: pandas/io/packers.py [hunk 2]
```diff
@@ -115,8 +115,12 @@ def writer(fh):
 
     path_or_buf = _stringify_path(path_or_buf)
     if isinstance(path_or_buf, str):
-        with open(path_or_buf, mode) as fh:
-            writer(fh)
+        try:
+            with open(path_or_buf, mode) as fh:
+                writer(fh)
+        except FileNotFoundError:
+            raise ValueError('path_or_buf is invalid or was not found')
```
Comment (WillAyd) [inline pandas/io/packers.py]: Hmm not to overcomplicated things but shouldn't we still raise a FileNotFoundError? ValueError is inconsistent with other io methods

Comment (jreback) [reply pandas/io/packers.py]: yeah I think just re-raising FileNotFoundError with the new message might be helpful @U09Kane

Comment (U09Kane) [reply pandas/io/packers.py]: Sure thing

**Comment by U09Kane** [comment]
I'm not really sure why it fails on Linting

**Comment by WillAyd** [comment]
We just merged in a PR to use black as a code formatting tool. If you install that and run locally against pandas/io/packers.py and pandas/tests/io/test_packers.py I think should resolve

File: pandas/io/packers.py [hunk 3]
```diff
@@ -139,8 +139,12 @@ def writer(fh):
 
     path_or_buf = _stringify_path(path_or_buf)
     if isinstance(path_or_buf, str):
-        with open(path_or_buf, mode) as fh:
-            writer(fh)
+        try:
+            with open(path_or_buf, mode) as fh:
+                writer(fh)
+        except FileNotFoundError as error:
+            error.strerror = "File b'{}' does not exist".format(path_or_buf)
```
Comment (jreback) [inline pandas/io/packers.py]: hmm just do a 

```raise FileNotFoundError(msg)```

its prob ok generally to mutate the exception object, but we don't do this in pandas, so for consistency pls change.

File: pandas/io/packers.py [hunk 4]
```diff
@@ -204,13 +208,11 @@ def read(fh):
...
         try:
-            exists = os.path.exists(path_or_buf)
-        except (TypeError, ValueError):
-            exists = False
-
-        if exists:
             with open(path_or_buf, "rb") as fh:
                 return read(fh)
+        except FileNotFoundError as error:
+            error.strerror = "File b'{}' does not exist".format(path_or_buf)
```
Comment (jreback) [inline pandas/io/packers.py]: same

File: pandas/tests/io/test_packers.py [hunk 5]
```diff
@@ -165,12 +165,18 @@ def __init__(self):
...
 
         msg = "Invalid file path or buffer object type: <class '{}'>"
+        invalid_path = os.path.join("nonexistent_dir", "df.msgpack")
         with pytest.raises(ValueError, match=msg.format("NoneType")):
             read_msgpack(path_or_buf=None)
         with pytest.raises(ValueError, match=msg.format("dict")):
             read_msgpack(path_or_buf={})
         with pytest.raises(ValueError, match=msg.format(r".*\.A")):
             read_msgpack(path_or_buf=A())
+        with pytest.raises(FileNotFoundError):
```
Comment (jreback) [inline pandas/tests/io/test_packers.py]: can you do a match on the error message

**Review by jreback — CHANGES_REQUESTED** [review]
some comments; ping on green.

File: pandas/tests/io/test_packers.py [hunk 6]
```diff
@@ -165,12 +165,16 @@ def __init__(self):
                 self.read = 0
 
         msg = "Invalid file path or buffer object type: <class '{}'>"
+        invalid_path = os.path.join("nonexistent_dir", "df.msgpack")
+        path_msg = "File b'{}' does not exist".format(invalid_path)
```
Comment (jreback) [inline pandas/tests/io/test_packers.py]: just use a regex for 'does not exist' I think this is not matching on windows becase of the escaped pathname.

**Comment by U09Kane** [comment]
Hopefully, everything is good now.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks sticking it out @U09Kane 

nice patch!

**Comment by U09Kane** [comment]
Thanks for the feedback

**Pull request merged**